clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/deploy/ops/actions/UpdateAutoScalingGroup.java

```diff
@@ -60,19 +74,142 @@ public Result apply(
+              Optional.ofNullable(description.getOnDemandPercentageAboveBaseCapacity())
+                  .orElse(existingDist.getOnDemandPercentageAboveBaseCapacity()),
+              Optional.ofNullable(description.getSpotAllocationStrategy())
+                  .orElse(existingDist.getSpotAllocationStrategy()),
+              Optional.ofNullable(description.getSpotInstancePools())
+                  .orElse(existingDist.getSpotInstancePools()),
+              Optional.ofNullable(description.getSpotPrice())
+                  .orElse(existingDist.getSpotMaxPrice()));
+
+      updateReq.withMixedInstancesPolicy(mip);
+    } else {
+      if (command.isReqToUpgradeAsgToMixedInstancesPolicy) {
+        // convert launch template to MIP
+        final MixedInstancesPolicy mip =
+            getMixedInstancesPolicy(
+                ltId,
+                ltVersion,
+                command.launchTemplateOverrides,
+                description.getOnDemandAllocationStrategy(),
+                description.getOnDemandBaseCapacity(),
+                description.getOnDemandPercentageAboveBaseCapacity(),
+                description.getSpotAllocationStrategy(),
+                description.getSpotInstancePools(),
+                description.getSpotPrice());
+
+        updateReq.withMixedInstancesPolicy(mip);
+      } else {
+        updateReq.withLaunchTemplate(
+            new LaunchTemplateSpecification().withLaunchTemplateId(ltId).withVersion(ltVersion));
+      }
+    }
+
+    try {
+      regionScopedProvider.getAutoScaling().updateAutoScalingGroup(updateReq);
+    } catch (Exception e) {
+      StringBuilder exceptionMsg =
+          new StringBuilder(
+              String.format("Failed to update server group %s.", description.getAsgName()));
+      if (StringUtils.isNotBlank(e.getMessage())) {
+        exceptionMsg.append(String.format("Error: %s\n", e.getMessage()));
+      }
+
+      try {
+        // Clean up newly created launch template version by the Saga Flow, in order to keep the
+        // latest version unaltered.
+        // This step is required because only the default and latest launch template versions for a
+        // launch template are cached.
+        // Not cleaning up will result in Internal Server Error for Clouddriver API requests and
+        // subsequent Deck errors.
+        if (command.getNewLaunchTemplateVersionNumber() != null) {
+          saga.log("[SAGA_ACTION] Cleaning up to keep the operation atomic.");
+          cleanUpOnFailure(
+              regionScopedProvider.getLaunchTemplateService(),
+              command.getLaunchTemplateVersion().getLaunchTemplateId(),
+              command.getNewLaunchTemplateVersionNumber());
+        }
+      } catch (Exception ex) {
+        exceptionMsg.append(
+            "Failed to clean up launch template version! Error: " + ex.getMessage());
+      }
+      throw new LaunchTemplateException(exceptionMsg.toString(), e);
```

Comment:
A sample error message in task API response for the scenario where:
- updateAutoScalingGroup fails 
- clean up of newly created launch template version is attempted but that also fails

Made the error message clearer to indicate 2 failures. 

```
{
   "key":"kato.tasks",
   "value":[
      {
         "exception":{
            "retryable":true,
            "cause":"Update Failed!!!",
            "details":[
               {
                  "message":"Update Failed!!!"
               },
               {
                  "retryable":true,
                  "userMessage":"null",
                  "message":"Failed to update server group hello-lt_pr-v004.Error: Update Failed!!!\nFailed to clean up launch template version! Error: Failed to delete launch template version 14 for launch template ID lt-077b0c6cd0a07309b because of launchTemplateVersionDoesNotExist",
                  "additionalAttributes":{
                     
                  }
               }
            ],
            "message":"Failed to update server group hello-lt_pr-v004.Error: Update Failed!!!\nFailed to clean up launch template version! Error: Failed to delete launch template version 14 for launch template ID lt-077b0c6cd0a07309b because of launchTemplateVersionDoesNotExist",
            "type":"EXCEPTION",
            "operation":"ModifyServerGroupLaunchTemplateAtomicOperation"
         },
         "resultObjects":[
            {
               "retryable":true,
               "cause":"Update Failed!!!",
               "details":[
                  {
                     "message":"Update Failed!!!"
                  },
                  {
                     "retryable":true,
                     "userMessage":"null",
                     "message":"Failed to update server group hello-lt_pr-v004.Error: Update Failed!!!\nFailed to clean up launch template version! Error: Failed to delete launch template version 14 for launch template ID lt-077b0c6cd0a07309b because of launchTemplateVersionDoesNotExist",
                     "additionalAttributes":{
                        
                     }
                  }
               ],
               "message":"Failed to update server group hello-lt_pr-v004.Error: Update Failed!!!\nFailed to clean up launch template version! Error: Failed to delete launch template version 14 for launch template ID lt-077b0c6cd0a07309b because of launchTemplateVersionDoesNotExist",
               "type":"EXCEPTION",
               "operation":"ModifyServerGroupLaunchTemplateAtomicOperation"
            }
         ],
         "id":"6f73a2ab-cabe-48ab-9d19-5e972aa8186d",
         "history":[
            {
               "phase":"ORCHESTRATION",
               "status":"Initializing Orchestration Task"
            },
            {
               "phase":"ORCHESTRATION",
               "status":"Processing op: ModifyServerGroupLaunchTemplateAtomicOperation"
            },
            {
               "phase":"ORCHESTRATION",
               "status":"Orchestration failed: ModifyServerGroupLaunchTemplateAtomicOperation | LaunchTemplateException: [Failed to update server group hello-lt_pr-v004.Error: Update Failed!!!\nFailed to clean up launch template version! Error: Failed to delete launch template version 14 for launch template ID lt-077b0c6cd0a07309b because of launchTemplateVersionDoesNotExist]"
            }
         ],
         "status":{
            "retryable":true,
            "completed":true,
            "failed":true
         }
      }
   ]
}
```

Reply:
Ref: https://github.com/spinnaker/clouddriver/pull/5367#discussion_r638110836
Thanks for running through the scenario. Sounds legit to me. I would like for us to revisit this section in a separate PR (to avoid big changes).  It would have been nice if the api supported updating an `asg` an inline `withTemplateVersion` 😄

Reply:
Sure, would you like to discuss concerns offline or suggest more changes to this Saga step? 
I would definitely like to get your concerns addressed, if any.